### PR TITLE
Journal file support

### PR DESCRIPTION
**Pull request by aluzzardi**
Add support for `_EXPERIMENTAL_DAGGER_JOURNAL`. This will dump raw
events to a log file, useful for post analysis.

Currently this will only be used internally and is subject to removal.

Related: #4248

### CODE REVIEW

**Review by vito — APPROVED** [review]
lgtm!

File: engine/engine.go
Comment (sipsma) [inline engine/engine.go:184]: nit: pre-existing issue but I think the warning should be sent to startOpts.LogOutput instead of stderr

**Review by sipsma — APPROVED** [review]

File: engine/engine.go
Comment (aluzzardi) [reply engine/engine.go]: Done

**Comment by gerhard** [comment]
I am just about to open a PR which will confirm that this change works as expected. FTR: https://github.com/dagger/tools/pull/29#issuecomment-1373878073

**Comment by gerhard** [comment]
I don't think that this works as expected.

Given the following diffs:

```diff
diff --git a/internal/mage/sdk/nodejs.go b/internal/mage/sdk/nodejs.go
index bc0c8412..3a73f849 100644
--- a/internal/mage/sdk/nodejs.go
+++ b/internal/mage/sdk/nodejs.go
@@ -32,6 +32,7 @@ func (t Nodejs) Lint(ctx context.Context) error {

        eg.Go(func() error {
                _, err = base.
+                       WithExec([]string{"env"}).
                        WithExec([]string{"yarn", "lint"}).
                        ExitCode(gctx)
                return err
@@ -70,7 +71,7 @@ func (t Nodejs) Test(ctx context.Context) error {
        _, err = nodeJsBase(c).
                WithMountedDirectory("/root/.docker", util.HostDockerDir(c)).
                WithExec([]string{"yarn", "test"}).
-               ExitCode(ctx)
+               File("/engine.log").Export(ctx, "engine.log")
        return err
 }
```

```diff
❯ git diff internal/mage/util/util.go
diff --git a/internal/mage/util/util.go b/internal/mage/util/util.go
index 288e2c09..95ea8f65 100644
--- a/internal/mage/util/util.go
+++ b/internal/mage/util/util.go
@@ -62,11 +62,12 @@ func RepositoryGoCodeOnly(c *dagger.Client) *dagger.Directory {
        })
 }

func WithDevEngine(c *dagger.Client, ctr *dagger.Container) *dagger.Container {
        // the cli bin is statically linked, can just mount it in anywhere
        dockerCli := c.Container().From("docker:cli").File("/usr/local/bin/docker")

        cliBinPath := "/.dagger-cli"
+       engineLog := "/engine.log"
        return ctr.
                // Mount in the docker cli + socket, this will be used to connect to the dev engine
                // container
@@ -80,7 +81,9 @@ func WithDevEngine(c *dagger.Client, ctr *dagger.Container) *dagger.Container {
                WithMountedFile(cliBinPath, DaggerBinary(c)).
                // Point the SDKs to use the dev engine via these env vars
                WithEnvVariable("_EXPERIMENTAL_DAGGER_CLI_BIN", cliBinPath).
-               WithEnvVariable("_EXPERIMENTAL_DAGGER_RUNNER_HOST", "docker-container://"+TestContainerName)
+               WithEnvVariable("_EXPERIMENTAL_DAGGER_RUNNER_HOST", "docker-container://"+TestContainerName).
+               WithEnvVariable("_EXPERIMENTAL_DAGGER_JOURNAL", engineLog)
+
 }

 func goBase(c *dagger.Client) *dagger.Container {
```

The `engine.log` file produced for `./hack/make sdk:nodejs:test` is empty:

image

I have also noticed that only a subset of these tests run in the dev engine (the ones under `NodeJS SDK Connect`):

image

I have also noticed that `yarn test` runs in the stable dev engine, specifically `dagger-engine-9ad18aca71b1b1ba` which currently uses `ghcr.io/dagger/engine:v0.3.7`.

While digging into this, I noticed that we use `dagger.io/dagger v0.4.2` in our tests. This means that we provision the engine in which our pipelines run by default differently than we do in the `main` branch. We are currently using the behaviour before https://github.com/dagger/dagger/pull/4194 was merged, specifically https://github.com/dagger/dagger/blob/sdk/go/v0.4.2/sdk/go/internal/engineconn/dockerprovision/image.go#L35-L146 . This is somewhat confusing, but more importantly it means that we cannot use this journal file feature **in the engine where the majority of our pipelines run**.

I think that all our CI workflows should use the dev SDK - `replace go.dagger.io/dagger => ../../sdk/go` in `internal/mage/go.mod` is my suggestion - and connect to the dev engine by default. In specific cases, we should use the latest released engine version, but not by default. This is a diagram of what I am proposing:

```mermaid
flowchart LR
    make(./hack/make sdk:node:lint)
    sdk_node_test(go run main.go -w ../.. sdk:node:test)
    engine_dev(go run main.go -w ../.. engine:dev)

    make --> engine_dev --> engine_dev_localhost --> test-dagger-engine
    engine_dev -.- sdk_node_test
    
    src["./sdk/nodejs"] --> |COPY| workdir

    subgraph Docker
        subgraph dagger-engine-SHA:v0.3.7
            engine_dev_localhost[/BUILD A DEV ENGINE/]
        end
        subgraph test-dagger-engine
            workdir["/workdir"]
            lint[yarn lint]
        end
    end

    sdk_node_test --> |_EXPERIMENTAL_DAGGER_RUNNER_HOST| test-dagger-engine
```

This is what is happening today:

```mermaid
flowchart LR
    make(./hack/make sdk:node:lint)
    sdk_node_test(go run main.go -w ../.. sdk:node:test)
    engine_dev(go run main.go -w ../.. engine:dev)

    make --> engine_dev --> engine_dev_localhost --> test-dagger-engine
    engine_dev -.- sdk_node_test
    
    src["./sdk/nodejs"] --> |COPY| workdir

    subgraph Docker
        subgraph dagger-engine-SHA:v0.3.7
            workdir["/workdir"]
            lint[yarn lint]
            engine_dev_localhost[/BUILD A DEV ENGINE/]
        end
        subgraph test-dagger-engine
            connect_via[[_EXPERIMENTAL_DAGGER_RUNNER_HOST]]
        end
    end

    sdk_node_test --> dagger-engine-SHA:v0.3.7
```

**Comment by sipsma** [comment]
> I think that all our CI workflows should use the dev SDK - `replace go.dagger.io/dagger => ../../sdk/go in internal/mage/go.mod` is my suggestion - and connect to the dev engine by default. 

Use of a go mod replace like that will create catch-22s in cases where we are making a backwards incompatible change to how `dagger session` communicates w/ the runner. We won't have a dev engine already running, so we need to build it, but in order to build it we will be using the Go SDK, which in this case is incompatible with the previous stable engine, so we will be stuck. In a previous state of our CI this catch-22 was possible and actually happened [here](https://github.com/dagger/dagger/pull/3983#issue-1462614315).

Other possible solutions (discussed some of these w/ @aluzzardi earlier today too):
1. The "stable" engine doesn't have to be an official release. Once the journal file feature is merged into `main` then we can update `internal/mage/go.mod` to point to that.
   * We'll also need to update `./hack/*` scripts to use `_EXPERIMENTAL_DAGGER_RUNNER_HOST` and download the right CLI for use w/ `_EXPERIMENTAL_DAGGER_CLI_BIN`, which is a bit annoying but not the utter end of the world. Can also simplify this part a bit by updating the engine images to push to tags for each git commit (rather than just `:main`) and some other tweaks.
1. Split out the code so that the stable SDK+engine are used only for building+running the dev engine; after that all other steps use the dev engine.
   * This is totally possible but when discussed previously ([buried somewhere in the threads here](https://github.com/dagger/dagger/pull/3755)) it was causing headaches due to the fact that it meant the codebase for bootstrapping the dev engine had to be completely separate from the rest of the ci codebase, which would result in a bunch of hard-to-maintain duplication.

Overall, I don't love it but I think `1.` above might be the best. If we put in the effort to make it easier for non-released SDKs automatically use the correct CLI+runner image, that will make it easier and technically benefit users too since it'll be easier for them to use the Go SDK from unreleased versions rather than always waiting for a new release.

**Comment by gerhard** [comment]
`2.` seems to me that is easiest to understand: we always run against the current engine, the one that corresponds to a specific SHA in the CI, or local state in dev (including any uncommitted changes).

I cannot help thinking that having a stable engine to build a dev engine seems similar to having a single tool and using it for everything. Since we were able to build the first engine without using an existing engine, could we do the same for the dev engine?

**Comment by gerhard** [comment]
Having talked to @sipsma about this, the cleanest & easiest to understand approach requires us to move the dev engine building code into a separate directory with its own `go.mod` since we need to reference a stable Go SDK. We will use this stable Go SDK to auto-provision a stable engine which will be used to build a dev engine (like we do now in the context of `internal/mage` dir). Once we have build the dev engine, all commands that run through `./hack/make` will run inside this dev engine (which is exactly what we want). The only concern is that we will be duplicating code between these two directories (the new one, and the existing `internal/mage`). It's worth attempting to solve this again - things were more complicated when we first tried this out.

**Comment by aluzzardi** [comment]
#3755 has a bunch of context on this.

AFAIK there were a bunch of trade-offs in complexity for small gains, that's why we binned the initiative

**Pull request merged**
